fix: remove a custom metric with filter

jogeshwar01 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: #8629 

### Description:
Added logic for updating filters when custom metrics are removed.

Example -

Before:

![image](https://github.com/lightdash/lightdash/assets/85165953/1986a7a7-9a21-4391-bbed-7bc4d8563618)
![image](https://github.com/lightdash/lightdash/assets/85165953/cf8cbc56-fab8-4b30-8ac4-46582d0cdfac)

On removing:

![image](https://github.com/lightdash/lightdash/assets/85165953/6e9b9cd9-62f3-475d-9923-bef8353a3634)
![image](https://github.com/lightdash/lightdash/assets/85165953/6ae47ef2-7a92-4b46-9b60-3ad5f7aaa72f)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

owlas commented:
let's review by tomorrow if possible

ZeRego commented on packages/common/src/types/filter.ts:
```diff
@@ -199,6 +199,60 @@ export const updateFieldIdInFilters = (
+        filterGroup.or = processGroupItems(filterGroup.or);
+        if (filterGroup.or.length === 0) filterGroup = undefined;
+        else
+            filterGroup.or = filterGroup.or.filter((item) =>
+                isNotEmptyGroup(item),
+            );
```
Since this is a recursive function, do we need this `else`? wouldn't the previous line `if (filterGroup.or.length === 0) filterGroup = undefined;` remove all empty groups?

jogeshwar01 replied:
@ZeRego 
This case was mostly to filter out the cases when the `and`, `or` are not of length 0 but rather contain a empty nested `and`, `or` inside. I realized this might fail while writing test cases due to the way the parameters are updated locally inside a function. There would be problems setting it to undefined as it is passed as a reference. So I have moved the code for undefined separately.
It works fine now. Just wanted to know, do you think it shouldn't be this complicated?

ZeRego replied:
> Just wanted to know, do you think it shouldn't be this complicated?

Thanks for the changes and especially the tests ❤️  I still ended up refactoring the function as I felt it could be simplified even more. 
It got easier to read and test once I removed the mutability nature. Then I was able to consolidate all the functions into 1 and only check once for empty groups.

ZeRego commented on packages/common/src/types/filter.ts:
```diff
@@ -199,6 +199,60 @@ export const updateFieldIdInFilters = (
+            );
+    } else if (isAndFilterGroup(filterGroup)) {
+        filterGroup.and = processGroupItems(filterGroup.and);
+        if (filterGroup.and.length === 0) filterGroup = undefined;
+        else
```
We should add brakets to all conditions for better readability.

ZeRego commented on packages/common/src/types/filter.ts:
```diff
@@ -199,6 +199,60 @@ export const updateFieldIdInFilters = (
+    }
+};
+
+export const removeMetricFromFilters = (
```
Can we add some unit tests? 

should return remaining filters ( simple case + nested groups )
should return undefined when all filter rules where removed ( simple case + nested groups )

jogeshwar01 commented:
@ZeRego 
- Added unit tests for both the above mentioned cases.
- Added brackets to all conditions.
- Updated code to correctly return undefined in case of all filters being removed.
